Minimal implementation of Broadcast Tables

margdoc opened the pull request:
Broadcast tables are tables for which all statements are strongly consistent (linearizable), replicated to every node in the cluster and available as long as a majority of the cluster is available. If a user wants to store a “small” volume of metadata that is not modified “too often” but provides high resiliency against failures and strong consistency of operations, they can use broadcast tables.

The main goal of the broadcast tables project is to solve problems which need to be solved when we eventually implement general-purpose strongly consistent tables: designing the data structure for the Raft command, ensuring that the commands are idempotent, handling snapshots correctly, and so on.

In this MVP (Minimum Viable Product), statements are limited to simple SELECT and UPDATE operations on the built-in table. In the future, other statements and data types will be available but with this PR we can already work on features like idempotent commands or snapshotting.  Snapshotting is not handled yet which means that restarting a node or performing too many operations (which would cause a snapshot to be created) will give incorrect results.

In a follow-up, we plan to add end-to-end Jepsen tests (https://jepsen.io/). With this PR we can already simulate operations on lists and test linearizability in linear complexity. This can also test Scylla's implementation of persistent storage, failure detector, RPC, etc.

Design doc: https://docs.google.com/document/d/1m1IW320hXtsGulzSTSHXkfcBKaG5UlsxOpm6LN7vWOc/edit?usp=sharing

margdoc commented:
@kbr- @kostja @gleb-cloudius @alecco @gusev-p  please review

kbr- commented on raft/group0_tables/lang.hh:
```diff
@@ -0,0 +1,35 @@
+    bytes key;
+    bytes new_value;
+    std::optional<bytes_opt> value_condition;
+};
```
Note to reviewers: this was probably the simplest approach to implement a simple key-value store.

In the future, if we want to handle all of CQL and other frontends (Alternator etc.), we could do something like:
- for selects, store a `read_command`
- for conditional updates, store a `read_command` (to read the data needed to check the condition), and some kind of data structure that describes a program which walks over the fetched mutation and performs the checks. Something like cql3 restriction object but more low-level, operating on mutations; furthermore, a data structure which describes how to create a mutation
- for updates, store a `mutation`

But I think this simple kv-store is a good starter, because - as explained in the cover letter - it will let us work on implementing idempotency, snapshotting, and testing it with Jepsen or otherwise.

avikivity replied:
But what's the plan? Throw it away later?

kbr- replied:
The current language definition is a placeholder. We'll throw most of it away. Fortunately most of the work we want to do is independent of the language definition. Designing the language is only one of the tasks that need to be solved in this project. I didn't want to make it a blocker task that must be solved before any other tasks, hence we have this placeholder for the MVP, and we can now decide what to work on.

margdoc replied:
As suggested by @kbr-, it will be done in the future

margdoc replied:
Language definition is moved to the experimental subdirectory to emphasize that it is subject to change

kbr- commented on test/group0_tables/test_group0_tables.py:
```diff
@@ -0,0 +1,91 @@
+                elif condition_type == ConditionType.NEW_KEY:
+                    key = random_string()
+                    random_strings.append(key)
+                    value_condition = None
```
I'm worried that the probability of choosing a key that already exists in `kv_store` is quite low with this approach.
Each of these cases gives a high chance that we'll choose a new key (where by new I mean not present in `kv_store`)

I would change the set of ConditionTypes to something like `[NEW_KEY, EXISTING_KEY_PASS, EXISTING_KEY_FAIL]` with about 20% chance for `NEW_KEY` (unless of course `kv_store` is empty, then choose `NEW_KEY`). For `NEW_KEY` we would randomly choose from the set difference `random_strings - keys_in_kv_store` (I don't think it's necessary to generate any more random strings, 42 is enough), for `EXISTING_KEY_*` we would randomly choose from `keys_in_kv_store` (and if `keys_in_kv_store` covers all of `random_strings`, we would always choose `EXISTING_KEY_*`). Then `PASS` means we pass the condition (so `value_condition` takes `kv_store[key]`), `FAIL` means we don't pass (so `value_condition` takes something different from `kv_store[key]`). `FAIL` vs `PASS` chance could be 50-50.

Also, the test should be reproducible if possible. We could roll the random generator seed at the beginning, print it, then pass it to the random generator. If one wants to reproduce the sequence of operations, they can reuse this seed.

dorlaor commented:
This is super exciting! (I know it's MVP only and group 0 only, still).
Cheers

On Fri, Jul 29, 2022 at 3:44 PM Mikołaj Grzebieluch <
***@***.***> wrote:

> @kbr- <https://github.com/kbr-> @kostja <https://github.com/kostja>
> @gleb-cloudius <https://github.com/gleb-cloudius> @alecco
> <https://github.com/alecco> @gusev-p <https://github.com/gusev-p> please
> review
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/11164#issuecomment-1199235606>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANHURNK53TZ4IPIVESXB73VWPGZVANCNFSM55AUAFAQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

avikivity commented:
Please rename it away from group 0. First, we'll want a separate group for it so heavy activity here doesn't interfere with everything else (I'm fine with keeping it in group 0 while it's still experimental, but the name has to change). Second, group 0 is an internal term without meaning for users.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -954,6 +954,20 @@ schema_ptr system_keyspace::discovery() {
+            .with_null_sharder()
+            .build();
+    }();
+    return schema;
```
Why not allow any schema with CREATE TABLE?

kbr- replied:
Reducing the scope of the first PR. Using a hardcoded table was the fastest way to let us focus on implementing the semantics of DML statements.

Of course one of the follow-up tasks will be to implement create keyspace/table for group 0 / strongly consistent tables.

kbr- replied:
You'll have to specify that you want a group 0 table in the DDL, so it's an additional syntax.
We'll also have to choose things like the replication strategy in the code.

So handling group 0 DDLs is some work.

avikivity replied:
I think this _is_ a new replication strategy.

kbr- replied:
Probably yes. I don't want implementing a whole new replication strategy to be part of the first PR. Let's make it one of the follow-up tasks.

margdoc replied:
As suggested by @kbr-, it will be changed in the future

avikivity commented on idl/group0_tables_lang.idl.hh:
```diff
@@ -0,0 +1,27 @@
+struct query {
+    std::variant<select_query, update_query> q;
+};
+
```
How will we scale this to encompass all of CQL?

denesb replied:
Actually I think this should follow the internal C++ api both in language and supported operations, otherwise we might end up marrying this to a single front-end language.

kbr- replied:
@denesb what do you mean by "the internal C++ api"?

kbr- replied:
> How will we scale this to encompass all of CQL?

Answering this question is a big task which will have its own design and implementation stages.

avikivity replied:
I think we should use CQL and prepared statements. We already have a huge mechanism to represent CQL (CQL language), deserialize it (the parser), cache the parsing work (prepared statements cache) and manage a remote handle the the cache (prepared statement ID).

kbr- replied:
1. If we want to use prepared statements cache, we cannot put the CQL string inside the Raft command - that would require every replica to parse it on apply.
2. But we also cannot put a prepared statement ID there, because we have no guarantee that the cache is the same on every replica.

Maybe we could put something like (string, ID) pair, a replica would first try to fetch the statement from its local cache using the ID, if it's not present then parse the string.
But to make this work efficiently we would need to ensure that the function CQL string -> ID is deterministic.

kbr- replied:
But I also don't want to put anything CQL specific inside the Raft commands because I want to make it possible to support other frontends.

We'll probably want to use the storage_proxy data structures in the end: `read_command`, `mutation` and so on. Both CQL and Alternator use them so at least those frontends would be covered.

avikivity replied:
The prepared statement ID is a hash of the CQL string, so that part isn't a problem.

There is a problem if the recipient doesn't have the CQL string, but it can just respond with a rejection, and the sender can retry with the CQL string included.

avikivity replied:
The other frontends can standardize on CQL as their language. Is CQL the perfect language? No. Is the language you're sketching on a napkin the perfect language? Also no.

kbr- replied:
> The other frontends can standardize on CQL as their language.

@nyh or @psarna could you give your input here? Would it be possible to compile Alternator to CQL?

kbr- replied:
> Is the language you're sketching on a napkin the perfect language? Also no.

This language is internal and we can change it as we like it or need it. CQL, not so much.

Also, I'm proposing we try to reuse the data structures from `storage_proxy`.

kbr- replied:
The piece of code we're currently discussing is the definition of the data structure which is put inside a Raft command, the thing that is replicated to every replica. Since you commented on it, I understood that you want to put the whole CQL statement inside (or equivalent of it).

kbr- replied:
> Then it generates a read_command for reads or a mutation for writes. Replicas don't need the full statement.

So you're basically agreeing that we want to use the `storage_proxy` data structures (`read_command`, `mutation`) for the Raft commands.

Now the question remains, what do we put there for conditional statements.

denesb replied:
> @denesb what do you mean by "the internal C++ api"?

Basically the `storage_proxy` API for reads and writes: `query()` and `mutate()`.

denesb replied:
> So you're basically agreeing that we want to use the `storage_proxy` data structures (`read_command`, `mutation`) for the Raft commands.

This is basically what I meant by our "C++ api".

> Now the question remains, what do we put there for conditional statements.

A conditional statement is basically an additional read (`read_command`) and a condition expression.

psarna replied:
> > The other frontends can standardize on CQL as their language.
> 
> @nyh or @psarna could you give your input here? Would it be possible to compile Alternator to CQL?

Compiling to a legal CQL query which can be parsed and processed by Scylla, no, not trivially - otherwise we wouldn't bother with coding the whole frontend layer and we'd just translate everything as is to CQL. But, especially with the new expression-based system, it's most likely possible to translate alternator requests to something that looks like CQL and back. But then, if you're considering serializing requests as strings, why not just store alternator requests, which are JSON, ergo strings? We do not allow accessing the same table from both alternator and CQL, so you might just as well store alternator requests as is.

avikivity replied:
True, we could forward alternator requests.

A downside is that is changes alternator from being a front-end to being involved in mid-layer processing.

avikivity replied:
I had some ideas about recovering prepared statements from non-prepared CQL and alternator requests:

1. Parse CQL/JSON and remember the locations of literals
2. Hash the input text, excluding the literals
3. If the the same hash recurs, create a prepared statement out of the input-text-excluding-literals-with-bind-markers-added (CQL) or a translation of the query to CQL (alternator)
4. add the text-excluding-literals to a recognition automaton (e.g. regex) to recognize it and capture the literals
5. pass queries first via the automaton to see if the match such an extracted prepared statement


The value for CQL is dubious since it encourages bad behavior.

But for alternator, if parsing is sufficiently expensive, perhaps worthwhile.

margdoc replied:
We discussed this and put it off for later when we decide how the command will look like.

avikivity commented on idl/group0_tables_lang.idl.hh:
```diff
@@ -0,0 +1,27 @@
+
+struct update_query {
+    bytes key;
+    bytes new_value;
```
We use managed_bytes for values to avoid large allocations that are difficult for the allocator.

margdoc replied:
I've noticed that the serializer hasn't instantiation for managed_bytes. Should I write it?

kbr- replied:
`atomic_cell`s, which `mutation`s are built of, use `managed_bytes` everywhere (see atomic_cell.hh). But if `managed_bytes` doesn't have a serializer, how are they sent over the network?
So I looked into idl/mutation.idl.hh and saw that they use `bytes`. E.g.:
```
class live_cell stub [[writable]] {
    api::timestamp_type created_at;
    bytes value;
};
```

kbr- replied:
Ah, but it's a `stub`, and in idl-compiler.py you can find the following comment:
```
    The following specifiers are allowed in a class declaration:
     - `final` -- if a class is marked with this keyword it will not contain a
       size argument. Final classes cannot be extended by a future version, so
       it should be used with care.
     - `stub` -- no code will be generated for the class, it's only there for
       documentation.
```

so apparently `live_cell` must have a "manual" serializer definition somewhere.

kbr- replied:
I think you should look for an example of a type that uses `managed_bytes` and is sent over the network - like `mutation`. Then find out how this type is being (de)serialized, and use that approach for your struct.

avikivity commented on cql3/operation.hh:
```diff
@@ -86,6 +86,10 @@ public:
 
+    const std::optional<expr::expression>& get_expression() const {
+        return _e;
+    }
```
This works, but only by accident, because all of our operations happen to have on expression.

The serialization should be done in the sub-classes and the entry point should be a virtual function.

margdoc replied:
Now `operation` modifies `update_query` and `get_expression` is unnecessary. Operations are prepared as before, in `modification_statement` and in the broadcast tables query's case it updates `update_query`.

avikivity commented on cql3/query_processor.cc:
```diff
@@ -518,6 +524,21 @@ query_processor::execute_prepared_without_checking_exception_message(
+    } else [[likely]] {
+        co_return co_await statement.execute_without_checking_exception_message(qp, query_state, options);
+    }
+}
```
This pessimizes the common case by introducing a coroutine (=allocation) even when you don't need it. This should be a non-coroutine with just an if, calling the two branches (which may or may not be coroutines).

But, I don't think this is the correct place. We should have prepare() create new subclass of cql_statement with its own execute_without_checking_exception_message() which can then do its thing. This way we only have one dispatch point in the fast path, not two.

kbr- replied:
> This way we only have one dispatch point in the fast path, not two.

How so? We would dispatch on `prepare` to decide which of the subclasses to create, then a dispatch on `execute_without_checking_exception_message` - a virtual call. That's two dispatches.

avikivity replied:
prepare() is amortized over all of the uses of the statement. In a normal execution, it is not called.

kbr- replied:
>  We should have prepare() create new subclass of cql_statement with its own execute_without_checking_exception_message() which can then do its thing.

`cql_statement` has a whole tree of subclasses: `modification_statement`, `select_statement`, `batch_statement` and so on.

Do you suggest that we should have a mirror tree for strongly consistent tables? Like `strongly_consistent_modification_statement`, `strongly_consistent_select_statement` etc.?

Please elaborate.

If this is your proposal, then I think a better approach would be to simply have a `compile` virtual function in the existing hierarchy.

avikivity replied:
> > We should have prepare() create new subclass of cql_statement with its own execute_without_checking_exception_message() which can then do its thing.
> 
> `cql_statement` has a whole tree of subclasses: `modification_statement`, `select_statement`, `batch_statement` and so on.

In fact, there are several variations of select_statement, each corresponding to a different query plan (regular select, indexed, parallelized).

> 
> Do you suggest that we should have a mirror tree for strongly consistent tables? Like `strongly_consistent_modification_statement`, `strongly_consistent_select_statement` etc.?
> 
> Please elaborate.

Yes. They can share a lot of their internals. For example, SELECT will forward to the leader, which will then just reuse the eventually consistent select (perhaps with a read barrier too).

> 
> If this is your proposal, then I think a better approach would be to simply have a `compile` virtual function in the existing hierarchy.

compile() exists, it's called prepare().

kbr- replied:
> compile() exists, it's called prepare().

`prepare` creates an object in an inheritance hierarchy. We need a `compile` which produces a (de)serializable (to/from bytes) data structure which we can put inside a Raft command. Even if we have a separate inheritance hierarchy, there will be an internal `compile` somewhere.

avikivity replied:
prepare() can create an object with the compiled query (CQL or otherwise) who's execute_with_some_long_other_things() sends to raft. The compilation must occur ahead of time, we cannot compile each time a query is executed.

margdoc replied:
> prepare() can create an object with the compiled query (CQL or otherwise) who's execute_with_some_long_other_things() sends to raft.

Did I understand correctly that you propose to add new subclass of `cql_statement` whose `execute_without_checking_exception_message` method will
1. compile `query`
2. send `query` to raft

and leave executing queries in state machine as it is now?

kbr- replied:
I don't think we want to compile in `execute_without_checking_exception_message`. We want to compile ahead of time, in `prepare`, then `execute_...` will only send to raft.

avikivity replied:
Yes. Execution should attach run-time parameters to a handle to a pre-compiled query and send it along.

Calling something compile() is a sure way to get it rejected from functions called execute().

margdoc replied:
Moved compilation to `prepare()`. Executing pre-compiled queries with parameters will be added in the future.

avikivity commented on cql3/statements/modification_statement.hh:
```diff
@@ -256,6 +256,15 @@ public:
+    const std::vector<lw_shared_ptr<column_condition>>& get_regular_conditions() const {
+        return _regular_conditions;
+    }
+
```
I believe that if you have a specialized insert/update statement for group0 these won't be needed, because you'll be getting them during the prepare stage.

avikivity commented on raft/group0_tables/lang.cc:
```diff
@@ -0,0 +1,183 @@
+
+static
+query compile(const cql3::cql_statement& statement) {
+    if (auto s = dynamic_cast<const cql3::statements::primary_key_select_statement*>(&statement)) {
```
And you won't have these dynamic casts.

margdoc commented:
> Please rename it away from group 0. First, we'll want a separate group for it so heavy activity here doesn't interfere with everything else (I'm fine with keeping it in group 0 while it's still experimental, but the name has to change). Second, group 0 is an internal term without meaning for users.

@avikivity
Is "strongly consistent tables" good? Do you suggest any other name?

avikivity commented:
> > Please rename it away from group 0. First, we'll want a separate group for it so heavy activity here doesn't interfere with everything else (I'm fine with keeping it in group 0 while it's still experimental, but the name has to change). Second, group 0 is an internal term without meaning for users.
> 
> @avikivity Is "strongly consistent tables" good? Do you suggest any other name?

Strongly consistent tables is not selective enough. This is "strongly consistent tables that are replicated everywhere and don't scale across nodes or shards".

If we agree (do we?) that the syntax to enable them should be via the replication strategy, then we can do

    CREATE KEYSPACE ks2 WITH replication = { 'class': 'EverywhereReplicationStrategy', 'consistency': 'strong' }

or

    CREATE KEYSPACE ks2 WITH replication = { 'class': 'EverywhereReplicationStrategy' } AND 'consistency = 'strong'

or

    CREATE KEYSPACE ks2 WITH replication = { 'class': 'StronglyConsistentBroadcastStrategy' }


then we can derive the name from the syntax - strongly consistent everywhere tables, or broadcast tables.

kostja commented:
Next step for this patch IMO:
1) Add a feature bit, experimental feature
2) Commit a patch which implements ReplicatedEverywhere keyspaces, under this feature bit. 
3) Commit the basic grammar (akin to what is done here)
4) Develop the raft command representation that would be able to represent all condition types, i.e. extend the grammar to support other op types.

kbr- commented:
> Add a feature bit, experimental feature

See first commit.

> Commit a patch which implements ReplicatedEverywhere keyspaces, under this feature bit.

So we want to be able to create keyspaces in the very first version? No hardcoded table?
Note that we will still allow only a specific schema table to be created (a simple string->string kv store). So it seems like it's kind of wasted effort compared to using just hardcoded table (I'm talking about the very first version).

kostja commented on test/group0_tables/test_group0_tables.py:
```diff
@@ -0,0 +1,91 @@
+#
```
I'd begin with the basic syntax/limitations CQL tests before jumping onto the randomized tests train. Can I derive a view on a group0 table? Can I use a UDF inside a group0  table expression? Can I add a secondary key? We should cover these things with tests (and of course the main use case - conditional updates)

kbr- replied:
> Can I derive a view on a group0 table? Can I use a UDF inside a group0 table expression?

Ugh, far outside the scope of the project.

kostja replied:
These are must-have questions if we want to have it as a generally available feature. I don't mean we should support it - I mean we should address it in  the spec, document and test.

avikivity replied:
Agree it's outside the scope, but it should be relatively easy to implement (as a follow up, relax) since it can all be done in one transaction.

margdoc replied:
Added a comment about the language limitation in the test.

kbr- commented:
@margdoc summarizing our discussion in the office with @avikivity and @tgrabiec:
- the temporary intermediate language (with key+value+condition) is fine to get us unblocked, but we need to make it clear that it's experimental, e.g. we need to put the language definitions in idl/experimental/..., group0_tables/experimental/... etc.
- `bytes` is fine.

kbr- commented:
And we'll probably switch to a passive replication approach in the end, whether the leader/primary calculates mutations and we commit mutations to the log.

We'll need to think about a good implementation of passive replication where we can calculate the mutations quickly (without depending on earlier mutations to be committed).

margdoc commented:
v2:
* moved the compilation of strongly consistent queries to `prepare()`
* renamed `group0_tables` to `broadcast_tables`
* changed some coroutines to non-coroutines with continuations
* changed the sampling in `test_group0_kv_store`

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1940/

kbr- commented on service/raft/raft_group0_client.cc:
```diff
@@ -293,4 +294,36 @@ group0_command raft_group0_client::prepare_command(table_query query) {
+raft::broadcast_tables::query_result raft_group0_client::query_result_guard::get() const {
+    assert(_client->_results.contains(_query_id) && _client->_results[_query_id].has_value());
+
+    return std::move(*_client->_results[_query_id]);
```
3 map searches! (two in assert, then the last here).
Use `unordered_map::find()` to reduce it to a single search.

kbr- replied:
Also don't use `assert`, use `if`+`on_internal_error`.

kbr- commented on service/raft/raft_group0_client.cc:
```diff
@@ -293,4 +294,36 @@ group0_command raft_group0_client::prepare_command(table_query query) {
 
+raft_group0_client::query_result_guard::query_result_guard(utils::UUID query_id, raft_group0_client& client)
+    : _query_id{query_id}, _client{&client} {
+    _client->_results.emplace(_query_id, std::nullopt);
```
I recommend checking if `emplace` succeeded (i.e. there wasn't an entry with this ID yet) and performing `on_internal_error` if it didn't.

margdoc commented on cql3/statements/select_statement.cc:
```diff
@@ -873,6 +875,68 @@ primary_key_select_statement::primary_key_select_statement(schema_ptr schema, ui
+           selection.get_columns()[0]->name() == "value";
+}
+
+strongly_consistent_select_statement::strongly_consistent_select_statement(schema_ptr schema, uint32_t bound_terms,
```
That's because all implementations of the `select_statement` class are in `select_statement.*` files so I followed this convention. Should I move it?

kbr- replied:
In that case, the question is why `strongly_consistent_modification_statement` doesn't follow this convention.

Personally I'd prefer them to live in separate files, but the most important thing is that we make consistent choices.

avikivity replied:
We could create a subnamespace for this, if people want.

margdoc commented on cql3/statements/strongly_consistent_modification_statement.cc:
```diff
@@ -0,0 +1,147 @@
+    return false;
+}
+
+void strongly_consistent_modification_statement::add_update_for_key(mutation& m, const query::clustering_range& range, const update_parameters& params, const json_cache_opt& json_cache) const {
```
> What made you decide to make strongly_consistent_modification_statement a subclass of modification_statement in the first place, and not e.g. a subclass of cql_statement_opt_metadata (which is a much more generic interface, basically validate and execute)?

That's because the return type of `raw::update_statement::prepare_internal` is a pointer to `modification_statement`, thus if we want `strongly_consistent_modification_statement` to be a subclass of `cql_statement_opt_metadata`, the return type has to be changed (what I think is impossible) or we have to add a new subclass of `raw::cf_statement`.

kbr- replied:
I really dislike the design of `modification_statement`.
First you see this:
```
/*
 * Abstract parent class of individual modifications, i.e. INSERT, UPDATE and DELETE.
 */
class modification_statement : public cql_statement_opt_metadata {
```

you'd expect a simple interface with one or two virtual methods.

But then you read further and see all these concrete (and public) member functions, which are basically implementation details of the subclasses. So now when we want to add another modification statement which executes updates a bit differently, we're left with all this stuff that we don't need and we have to implement empty overrides of virtual members. This whole class is one big antipattern, doing "composition" by inheritance.

Anyway, enough with the ranting.

> the return type has to be changed (what I think is impossible)

Can you explain why you think it's impossible?

Does the caller of `prepare_internal` use the concrete type information? If they do - what is it used for? Maybe this could be refactored. The purpose of `prepare` should be to return an abstract statement type which offers a way to bind variables and to execute it. Indeed, `prepare` returns a `prepared_statement` which contains `shared_ptr<cql_statement>` inside and `cql_statement` is very abstract. I'm wondering why we would need to return anything more concrete at any point.

avikivity replied:
> I really dislike the design of `modification_statement`. First you see this:
> 
> ```
> /*
>  * Abstract parent class of individual modifications, i.e. INSERT, UPDATE and DELETE.
>  */
> class modification_statement : public cql_statement_opt_metadata {
> ```
> 
> you'd expect a simple interface with one or two virtual methods.
> 
> But then you read further and see all these concrete (and public) member functions, which are basically implementation details of the subclasses. So now when we want to add another modification statement which executes updates a bit differently, we're left with all this stuff that we don't need and we have to implement empty overrides of virtual members. This whole class is one big antipattern, doing "composition" by inheritance.

Agree with this criticism. Sharing implementation details should be done via helper functions, or via a further subclass.

> 
> Anyway, enough with the ranting.
> 
> > the return type has to be changed (what I think is impossible)
> 
> Can you explain why you think it's impossible?
> 
> Does the caller of `prepare_internal` use the concrete type information? If they do - what is it used for? Maybe this could be refactored. The purpose of `prepare` should be to return an abstract statement type which offers a way to bind variables and to execute it. Indeed, `prepare` returns a `prepared_statement` which contains `shared_ptr<cql_statement>` inside and `cql_statement` is very abstract. I'm wondering why we would need to return anything more concrete at any point.

It might be premature use of a covariant return type and likely can be degraded.

margdoc replied:
> Does the caller of prepare_internal use the concrete type information? If they do - what is it used for?

`prepare_internal` is called inside `modification_statement::prepare`.  ` modification_statement::prepare` clears the `ctx` cache (for non-conditional statements) after calling `prepare_internal`.

avikivity replied:
Maybe that can be moved to individual statement implementations.

kbr- replied:
This piece of code looks like a good candidate to be moved into this `prepare_conditions` helper, which is called by `prepare_internal` implementations where applicable.

But be careful: some subclasses call `process_where_clause` after `prepare_conditions`, and it appears that this piece of code requires both `process_where_clause` and `prepare_conditions` to finish first (it accesses `_restrictions`, which is calculated in `process_where_clause` AFAIK).

Perhaps we could move `prepare_conditions` after `process_where_clause` in those implementations? And state in the interface of `prepare_conditions` that `_restrictions` must be calculated by this point, or something like that. Would have to check if there is no dependence in the other direction already to see if that's possible.

If we were able to move `prepare_conditions` after `process_where_clause,` then we could move this piece of code at the end of `prepare_conditions`.

kbr- replied:
Next question is whether all of this is worth it.

Even if we manage to change the return type of `prepare_internal` to return a more abstract base class, there's still this part of `prepare_internal` happening after `create_statement()` call (`create_statement which you added), which depends on the concrete type returned by `create_statement`, which right now returns `modification_statement`. There are calls to `add_operation` and `prepare_conditions` and `process_where_clause`. I'm guessing our `strongly_consistent_modification_statement` needs some of this stuff too.

I see two approaches we can take:
Approach 1:
- refactor the part inside `prepare_internal` after `create_statement` into a virtual function, with one implementation for `strongly_consistent_modification_statement` and another implementation shared by the other statements
- the parts which both types of statements need would have to be refactored into helper objects/functions so there is no logic duplication
- note: this would also solve the `prepare_raft_comand` problem

Approach 2:
- ignore these design problems and live with the pain of subclassing `modification_statement`  and having empty implementations; `modification_statement` has a refactor coming for it eventually anyway :>

I would prefer approach 1, but you should investigate how much pain it would take to do the refactors, and how much refactoring is actually needed, and then decide for yourself (and convince us that it's too much pain).

avikivity replied:
If you do decide to refactor, it's a good idea to retire it early in a separate pull request.

margdoc replied:
`raw::modification_statement::prepare` is also called by `batch_statement` which stores a vector of `modification_statement`, thus we cannot simply change the return type of `prepare_internal`.

margdoc replied:
I would choose the second approach. `prepare_conditions` can't be simply moved after `process_where_clause`. `process_where_clause` validates the primary key based on prepared conditions. It can be moved from `process_where_clause`, but then we will probably end up with a new virtual function with specialization only for one subclass (`insert_json_statement` doesn't call `process_where_clause` and doesn't have the primary key validation).

margdoc replied:
`strongly_consistent_modification_statement` is now prepared by `modification_statement` and derives from `cql_statement_opt_metadata`. Preparation of `modification_statement` is reused: for every request, it is prepared and if it's broadcast tables query, it will prepare (`prepare_for_raft`) and return the raft command. `prepare_for_raft` is implemented only for `update_statement`, but it will be eventually (if we decide to choose active replication) implemented in others statements (for now, there is a default implementation that throws an exception).

margdoc commented:
v3:
* renamed `group0_kv_store` to `broadcast_kv_store`
* moved `strongly_consistent_select_statement` to the separate file
* `create_guard` -> `create_result_guard`
* fixed `query_result_guard`'s comment
* fixed usage of `std::unordered_map`
* moved `execute_broadcast_table_query` to `lang.cc`
* moved creating of `cql_transport::messages::result_message::bounce_to_shard` to `strongly_consistent_*.cc` files

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2079/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2085/

margdoc commented:
v4:
* changed the way how operations are prepared in raft commands
* moved preparation of `strongly_consistent_modification_statement` after preparation of `modification_statement`

margdoc commented on cql3/statements/update_statement.cc:
```diff
@@ -384,6 +399,8 @@ update_statement::prepare_internal(data_dictionary::database db, schema_ptr sche
     prepare_conditions(db, *schema, ctx, *stmt);
     stmt->process_where_clause(db, _where_clause, ctx);
+
+    stmt->prepare_raft_command();
```
Moved the preparation of `strongly_consistent_modification_statement` after the preparation of `modification_statement`.

kbr- commented on cql3/constants.hh:
```diff
@@ -44,6 +46,10 @@ public:
         }
+
+        virtual void prepare_for_raft(raft::broadcast_tables::update_query& query) const override {
+            query.new_value = expr::evaluate(*_e, query_options::DEFAULT).to_bytes();
```
IIUC this will be an error if `_e` contains a bind marker instead of a constant. Please leave a FIXME comment that this works only for constants and bind markers are unimplemented at this point.

Also, I'm guessing that we don't have access to `update_parameters` at this point (compared to `execute`) - they only appear at execution time, and they contain e.g. values for bind markers, right?

kbr- replied:
Also, could we move this implementation to some .cc file? I don't like the fact that we now include
```
#include "raft/broadcast_tables/experimental/lang.hh"
```
inside cql3/constants.hh, which is potentially included in many places.

There's already a cql3/constants.cc file - you could put the impl there.

margdoc replied:
> Please leave a FIXME 

Done

> Also, I'm guessing that we don't have access to update_parameters at this point (compared to execute) - they only appear at execution time, and they contain e.g. values for bind markers, right?

Right

> There's already a cql3/constants.cc file - you could put the impl there.

Done

kbr- commented on cql3/operation.cc:
```diff
@@ -357,4 +358,9 @@ operation::element_deletion::prepare(data_dictionary::database db, const sstring
 
+void
+operation::prepare_for_raft(raft::broadcast_tables::update_query&) const {
+    throw raft::broadcast_tables::unsupported_operation_error{};
```
Is there no easy way to provide some details at this place (e.g. maybe there's a way to pretty-print (or ugly-print) the operation that we attempted to execute)?

kbr- replied:
Also `prepare_for_raft` should eventually become a pure virtual function in `operation` and it should be implemented in every type of `operation`.

Please live a FIXME which tells us that, so the reader of the code knows this state is temporary (there is no inherent reason that we throw unsupported here other than it's WIP).

margdoc replied:
> Is there no easy way to provide some details at this place (e.g. maybe there's a way to pretty-print (or ugly-print) the operation that we attempted to execute)?

We can use the `_e` attribute ( `std::optional<expr::expression>`), but it will work
> only by accident, because all of our operations happen to have on expression.

or use a column definition, but then we won't get the information about a type of the operation.

margdoc replied:
> Please live a FIXME which tells us that, so the reader of the code knows this state is temporary (there is no inherent reason that we throw unsupported here other than it's WIP).

Done

kbr- commented on cql3/statements/modification_statement.cc:
```diff
@@ -439,13 +441,30 @@ modification_statement::process_where_clause(data_dictionary::database db, expr:
 namespace raw {
 
+::shared_ptr<cql_statement_opt_metadata>
+modification_statement::prepare_statement(data_dictionary::database db, prepare_context& ctx, cql_stats& stats) {
```
The number of different functions named `prepare_*` make my head hurt.

I would prefer this to be inlined in `prepare`, don't see much advantage to introducing yet another helper function (which is very simple and used in exactly one place).

But I'll leave this decision to you.

margdoc replied:
`modifiation_statement` has to be prepared and then it prepares `strongly_consistent_modifiation_statement`, which doesn't derive from `modifiation_statement`. Thus we can't reassign the `statement` variable (like you mentioned before) and we can't change variable's type (we use the `prepare_for_raft` method). I would leave it in a separate method.

kbr- replied:
Ah, so we would need two variables to inline it.
Meh, ok, separate method is fine.

kbr- commented on cql3/statements/modification_statement.cc:
```diff
@@ -439,13 +441,30 @@ modification_statement::process_where_clause(data_dictionary::database db, expr:
 }
 
+::shared_ptr<strongly_consistent_modification_statement>
+modification_statement::prepare_for_raft() const {
```
`prepare_for_raft` should eventually become a pure virtual function in `modification_statement` and it should be implemented in every type of `modification_statement`.

Please live a FIXME which tells us that, so the reader of the code knows this state is temporary (there is no inherent reason that we throw unsupported here other than it's WIP).

margdoc replied:
Done

kbr- commented on cql3/statements/update_statement.cc:
```diff
@@ -254,6 +257,92 @@ void insert_prepared_json_statement::execute_operations_for_key(mutation& m, con
+static
+void prepare_new_value(raft::broadcast_tables::update_query& query, const std::vector<::shared_ptr<operation>>& operations) {
+    if (operations.size() != 1) {
+        throw raft::broadcast_tables::unsupported_operation_error(fmt::format("only operation \"value = X\" is supported"));
```
if one tries to do:
`v1 = X, v2 = Y`
they will get an error
"only operation value = X is supported"
and get confused.

The clou here is that only a SINGLE operation (of any type) is supported right now. Please adjust the comment.

margdoc replied:
Changed this to `"only one operation is allowed and must be of the form \"value = X\""`.

kbr- reviewed (commented):
I like this design much better, left some minor comments.

kbr- commented on raft/broadcast_tables/experimental/lang.cc:
```diff
@@ -25,7 +27,99 @@ bool is_broadcast_table_statement(const sstring& keyspace, const sstring& column
+
+        // Read mutations
+        const auto [read_cmd, range] = prepare_read_command(proxy, *schema, q.key);
+        auto [rs, _] = co_await proxy.query_mutations_locally(schema, read_cmd, range, db::no_timeout);
```
Nit: you use `proxy` before `co_await` and `_proxy` after `co_await`. Kinda weird. Maybe use `_proxy` everywhere?

Also, I generally suggest a different approach: instead of capturing, take these stuff through regular function arguments, then `std::bind_front` immediately. Then I think it's harder to make a mistake. But it's up to you.

margdoc replied:
Thanks for the advice with `std::bind_front`, now it uses `std::bind_front ` instead of lambda capture.

kbr- commented on raft/broadcast_tables/experimental/lang.cc:
```diff
@@ -26,9 +26,57 @@ bool is_broadcast_table_statement(const sstring& keyspace, const sstring& column
     auto group0_cmd = group0_client.prepare_command(service::broadcast_table_query{query});
-    return group0_client.add_entry_unguarded(std::move(group0_cmd));
+    auto guard = group0_client.create_result_guard(group0_cmd.new_state_id);
+    return group0_client.add_entry_unguarded(std::move(group0_cmd)).then([guard = std::move(guard)] () {
```
Coroutines please

margdoc replied:
Done

kbr- commented on raft/broadcast_tables/experimental/lang.cc:
```diff
@@ -26,9 +26,57 @@ bool is_broadcast_table_statement(const sstring& keyspace, const sstring& column
 
-future<> execute(service::raft_group0_client& group0_client, const query& query) {
+::shared_ptr<cql_transport::messages::result_message>
+to_cql_result(raft::broadcast_tables::query_result result) {
```
The translation from lower layer to upper layer should happen in the upper layer; the lower layer should not be aware of the upper layer

kbr- replied:
Any includes from `cql3/*` should eventually disappear

kbr- commented on service/raft/raft_group0_client.cc:
```diff
@@ -372,4 +373,44 @@ db::system_keyspace& raft_group0_client::sys_ks() {
+    : _query_id{query_id}, _client{&client} {
+    auto it = _client->_results.emplace(_query_id, std::nullopt);
+    if (!it.second) {
+        on_internal_error(logger, "query_result_guard::query_result_guard: creating an entry for the result doesn't succeed");
```
"doesn't succeed" is a weird error.
Provide the specifics: there was an unexpected entry there
(how could there be an entry if we generate a fresh UUID?)

margdoc replied:
Changed to `"query_result_guard::query_result_guard: there is another query_result_guard alive with the same query_id"`

kbr- commented on service/raft/raft_group0_client.cc:
```diff
@@ -372,4 +373,44 @@ db::system_keyspace& raft_group0_client::sys_ks() {
 
+raft_group0_client::query_result_guard::query_result_guard(utils::UUID query_id, raft_group0_client& client)
+    : _query_id{query_id}, _client{&client} {
+    auto it = _client->_results.emplace(_query_id, std::nullopt);
```
`it` should be reserved for iterators, here it's not an iterator but a pair <iterator, bool>
I suggest:
`auto [_, emplaced] = ...`

kbr- commented on service/raft/raft_group0_client.cc:
```diff
@@ -372,4 +373,44 @@ db::system_keyspace& raft_group0_client::sys_ks() {
+    }
+}
+
+raft::broadcast_tables::query_result raft_group0_client::query_result_guard::get() const {
```
Is this function really logically `const`? We do `move` out of the guarded entry

margdoc replied:
You're right, I removed `const`.

kbr- commented on test/broadcast_tables/conftest.py:
```diff
@@ -0,0 +1,15 @@
+import pathlib
+import sys
+
+# Add test.pylib to the search path
```
Is this actually needed? Do we import from `pylib` anywhere?

margdoc replied:
You're right, I removed the `conftest.py` file.

kbr- commented:
I left minor comments only, I think the general shape looks good.
@avikivity please review. I'm hoping to get this merged by Friday.

margdoc commented on cql3/statements/strongly_consistent_modification_statement.cc:
```diff
@@ -0,0 +1,74 @@
+#include "cql3/operation.hh"
+#include "cql3/query_processor.hh"
+#include "raft/broadcast_tables/experimental/lang.hh"
+#include "seastar/core/future.hh"
```
Fixed

margdoc commented on cql3/statements/select_statement.hh:
```diff
@@ -18,6 +18,7 @@
 #include "index/secondary_index_manager.hh"
 #include "exceptions/exceptions.hh"
 #include "exceptions/coordinator_result.hh"
+#include "raft/broadcast_tables/experimental/lang.hh"
```
It isn't needed, I removed it.

margdoc commented on cql3/statements/select_statement.cc:
```diff
@@ -14,7 +14,10 @@
 #include "cql3/statements/prune_materialized_view_statement.hh"
+#include "cql3/statements/strongly_consistent_select_statement.hh"
 
+#include "cql3/stats.hh"
```
No, now it is deleted.

margdoc commented:
v5:
* fixed includes
* changed lambda capture to `std::bind_front`
* left FIXMEs
* moved the logic of parsing `query_result` to `cql3`
* improved error messages
* added checking if update statement attributes were passed
* `it` -> `[_, emplaced]`
* removed `const` qualifier from `raft_group0_client::query_result_guard::get()`
* changed `raft::broadcast_tables::execute` and `strongly_consistent_*_statement::execute_without_checking_exception_message` to coroutines
* moved `cql3::constants::setter::prepare_for_raft` to `.cc` file
* removed unnecessary `conftest.py` file

kbr- reviewed (approved):
LGTM

@avikivity please review

avikivity commented on raft/broadcast_tables/experimental/lang.hh:
```diff
@@ -0,0 +1,37 @@
+// If value_condition is nullopt, the update is unconditional.
+struct update_query {
+    bytes key;
+    bytes new_value;
```
managed_bytes, so we can support large values

kbr- replied:
Recall that we agreed in the office that we can go with `bytes` for now. Switching to `managed_bytes` is complex because there is no straightforward way to serialize them. We can try to do it as a follow-up, but we'll switch to a passive replication approach eventually anyway so I think `bytes` can stay for initial testing.

avikivity commented on raft/broadcast_tables/experimental/lang.hh:
```diff
@@ -0,0 +1,37 @@
+struct update_query {
+    bytes key;
+    bytes new_value;
+    std::optional<bytes_opt> value_condition;
```
The condition is that the value matches some old value, so a more accurate name is old_value or some such. I'd expect a variable named "condition" to contain an expression, not a value. Also bytes_opt -> managed_bytes_opt.

The evaluation of a NULL here should match nothing. So if the outer optional is engaged and the inner optional is disengaged. The query is not applied.

kbr- replied:
> The evaluation of a NULL here should match nothing. So if the outer optional is engaged and the inner optional is disengaged. The query is not applied.

But that's not how LWT works, IIRC. And we don't want to introduce different semantics.
In LWT when you do `IF v = null` the update will be applied iff the column was `null`.

avikivity commented on cql3/operation.cc:
```diff
@@ -18,6 +18,7 @@
 #include "types/list.hh"
 #include "types/set.hh"
 #include "types/user.hh"
+#include "raft/broadcast_tables/experimental/lang.hh"
```
This should not be under raft/. It's using raft but it's a lot more cql and database than raft.

It's okay to move it post-merge.

kbr- replied:
Maybe `service/raft/broadcast_tables/`

margdoc replied:
> It absolutely shouldn't.

Right, I will move it.

> How hard is it to move them now?

I think it won't be hard.

kbr- replied:
In that case, the namespaces should be fixed too. If you want to move it under service/, raft::broadcast_tables -> service::raft::broadcast_tables.

avikivity replied:
But it will pull cql into group0 code. I'd be much happier if it was in a different group. As it is, whereever we put it it will entangle two unrelated things, so we might as well keep it here.

gleb-cloudius replied:
Raft is an independent library. It is fine to entangle it with group0 while it uses group0, but not with entire raft. BTW `service/raft/` contains not only group0 code. Generic raft group management is also there. We can also put it into `service/broadcast_tables/` and pretend that raft usage is just an implementation detail.

kbr- replied:
> But it will pull cql into group0 code

Will it? broadcast tables module has no CQL specific code. The CQL specific parts live in cql3/statements/strongly_consistent_* files.

avikivity replied:
ok

avikivity commented on cql3/statements/modification_statement.cc:
```diff
@@ -439,13 +442,31 @@ modification_statement::process_where_clause(data_dictionary::database db, expr:
 
+::shared_ptr<strongly_consistent_modification_statement>
+modification_statement::prepare_for_raft() const {
+    // FIXME: implement for every type of `modification_statement`.
```
This should be "prepare_for_broadcast_tables". Raft is just a tool here.

margdoc replied:
Done.

avikivity commented on cql3/statements/modification_statement.cc:
```diff
@@ -439,13 +442,31 @@ modification_statement::process_where_clause(data_dictionary::database db, expr:
+    ::shared_ptr<cql3::statements::modification_statement> statement = prepare(db, ctx, stats);
+
+    if (db.get_config().check_experimental(db::experimental_features_t::feature::BROADCAST_TABLES) &&
+            raft::broadcast_tables::is_broadcast_table_statement(keyspace(), column_family())) [[unlikely]] {
```
Isn't the experimental check unneeded? is_broadcast_table_statement() will return false if it's not enabled.

Minor: [[unlikely]] is unneeded, since prepare is rare.

margdoc replied:
`is_broadcast_table_statement()` is meant to check only if the query should be executed on a broadcast table. I can change it to check also if this feature is enabled.

avikivity replied:
I thought the table doesn't exist if the feature isn't enabled.

margdoc replied:
Yes, you're right. I will fix it.

margdoc replied:
Removed experimental check and [[unlikely]].

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1734,6 +1736,21 @@ std::unique_ptr<prepared_statement> select_statement::prepare(data_dictionary::d
             std::move(prepared_attrs)
         );
+    } else if (db.get_config().check_experimental(db::experimental_features_t::feature::BROADCAST_TABLES) &&
+               raft::broadcast_tables::is_broadcast_table_statement(keyspace(), column_family())) [[unlikely]] {
```
Again the experimental check and unlikely can be dropped.

margdoc replied:
Done.

avikivity commented on cql3/statements/strongly_consistent_select_statement.cc:
```diff
@@ -0,0 +1,90 @@
+
+    if (!column || column->col->kind != column_kind::partition_key ||
+        !value || key_restriction->op != cql3::expr::oper_t::EQ) {
+        throw raft::broadcast_tables::unsupported_operation_error(fmt::format("key restriction: {}", *key_restriction));
```
This should be caught in the prepare stage, yes?

So this should be on_internal_error().

avikivity replied:
Ah, I was confused since you're fetching the value here. Values are fetched during the execution stage since not all values are constant.

avikivity commented on cql3/statements/strongly_consistent_select_statement.cc:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    const auto* column = cql3::expr::as_if<cql3::expr::column_value>(&key_restriction->lhs);
+    const auto* value = cql3::expr::as_if<cql3::expr::constant>(&key_restriction->rhs);
```
What about bind variables?

kbr- replied:
I'd suggest leaving a FIXME

margdoc replied:
Added FIXME

avikivity commented on cql3/statements/strongly_consistent_select_statement.cc:
```diff
@@ -0,0 +1,90 @@
+        throw raft::broadcast_tables::unsupported_operation_error(fmt::format("key restriction: {}", *key_restriction));
+    }
+
+    return to_bytes(value->view());
```
If you call expr::evaluate() here, then bind variables will work. We can defer it to a later stage (though it's quite important if you want to stress raft).

margdoc replied:
For evaluating bind variables we need query options, which are available during query execution, not in the preparation stage. This code is executed in `prepare()`, thus we can't bind values here. 

We can add bind variables by changing `key/value/value_condition` type to `variant<bytes, bind_marker>` and then bind it in `execute()`. It can be done later in the follow-up.

avikivity replied:
Ok. It's generally important to have good separation between preparation and execution (an exception is constant subexpression elimination which is execution that is done during preparation).

avikivity commented on raft/broadcast_tables/experimental/lang.cc:
```diff
@@ -22,7 +24,95 @@ bool is_broadcast_table_statement(const sstring& keyspace, const sstring& column
+    const query& query,
+    utils::UUID cmd_id) {
+    return std::visit(make_visitor (
+    std::bind_front([] (service::storage_proxy& proxy, const raft::broadcast_tables::select_query& q) -> future<> {
```
Need some indents here.

kbr- replied:
We usually don't indent the arguments to `make_visitor`, from a quick glance through the codebase

avikivity replied:
Well it looks bad. std::bind_front is a continuation of the return statement, without indents it will look like a separate thing to a casual reader.

margdoc replied:
Added indent.

avikivity commented on raft/broadcast_tables/experimental/lang.cc:
```diff
@@ -22,7 +24,95 @@ bool is_broadcast_table_statement(const sstring& keyspace, const sstring& column
+            new_mutation.set_clustered_cell(clustering_key::make_empty(), "value", std::move(value), ts);
+            co_await proxy.mutate_locally(new_mutation, {}, {});
+        }
+    }, std::ref(proxy), cmd_id)), query.q);
```
I guess you're using bind_front because the outer function isn't a coroutine and the inner functions are coroutines, and you want to avoid losing the captures? Better make a comment about it.

margdoc replied:
> I guess you're using bind_front because the outer function isn't a coroutine and the inner functions are coroutines, and you want to avoid losing the captures?

Yes, added a comment.

avikivity commented on service/raft/raft_group0_client.cc:
```diff
@@ -13,6 +13,9 @@
 #include "schema_mutations.hh"
+#include "raft/broadcast_tables/experimental/lang.hh"
+#include "idl/experimental/broadcast_tables_lang.dist.hh"
+#include "idl/experimental/broadcast_tables_lang.dist.impl.hh"
```
I'm unhappy about mixing in all this stuff into group0, but okay as it's experimental (and destined to be replaced, not graduated into production). I retract my comment about moving it ourside raft/, it's too deeply entangled.

kbr- replied:
We should move outside raft/, because raft/ contains the library and only the library; the users of the library should be somewhere else. All our users reside in service/raft/ so I think that's the proper place

avikivity replied:
I retract my retraction.

margdoc replied:
Moved from `raft/broadcast_tables` to `service/broadcast_tables/`.

avikivity commented:
Very nice, I had just minor comments.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2143/

avikivity commented:
> I think we're good. I'll merge once CI passes (unless there's an objection from @avikivity). 

No.

> If any new minor comments appear, I'd suggest leaving them for a follow-up.

Let's get support for bind variables in.

kbr- commented:
Interesting - next promotion failed (https://jenkins.scylladb.com/job/scylla-master/job/next/5257/consoleText) because of:
```
[2022-09-09T11:15:09.866Z] ------------------------------------------------------------------------------
[2022-09-09T11:15:09.866Z] The following test(s) have failed: broadcast_tables/test_broadcast_tables broadcast_tables/test_broadcast_tables broadcast_tables/test_broadcast_tables
[2022-09-09T11:15:09.866Z] Output of pytest --host=127.122.212.1 -s --log-level=DEBUG -o junit_family=xunit2 --junit-xml=/jenkins/workspace/scylla-master/next/scylla/testlog/x86_64/release/xml/test_broadcast_tables.6232.xunit.xml test/broadcast_tables/test_broadcast_tables.py:
[2022-09-09T11:15:09.866Z] === TEST.PY STARTING TEST test_broadcast_tables.6232 ===
[2022-09-09T11:15:09.866Z] export UBSAN_OPTIONS='halt_on_error=1:abort_on_error=1:suppressions=/jenkins/workspace/scylla-master/next/scylla/ubsan-suppressions.supp'
[2022-09-09T11:15:09.866Z] export ASAN_OPTIONS='disable_coredump=0:abort_on_error=1:detect_stack_use_after_return=1'
[2022-09-09T11:15:09.866Z] pytest --host=127.122.212.1 -s --log-level=DEBUG -o junit_family=xunit2 --junit-xml=/jenkins/workspace/scylla-master/next/scylla/testlog/x86_64/release/xml/test_broadcast_tables.6232.xunit.xml test/broadcast_tables/test_broadcast_tables.py
[2022-09-09T11:15:09.866Z] === TEST.PY TEST test_broadcast_tables.6232 OUTPUT ===
[2022-09-09T11:15:09.866Z] ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
[2022-09-09T11:15:09.866Z] pytest: error: unrecognized arguments: --host=127.122.212.1
[2022-09-09T11:15:09.866Z]   inifile: /jenkins/workspace/scylla-master/next/scylla/test/broadcast_tables/pytest.ini
[2022-09-09T11:15:09.866Z]   rootdir: /jenkins/workspace/scylla-master/next/scylla/test/broadcast_tables
[2022-09-09T11:15:09.866Z] 
[2022-09-09T11:15:09.866Z] === TEST.PY SUMMARY START ===
[2022-09-09T11:15:09.866Z] Test exited with code 4
[2022-09-09T11:15:09.866Z] 
[2022-09-09T11:15:09.866Z] === TEST.PY SUMMARY END ===
```

I don't get it.
```
[2022-09-09T11:15:09.866Z] === TEST.PY TEST test_broadcast_tables.6232 OUTPUT ===
[2022-09-09T11:15:09.866Z] ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
```
But other pytests have passed.
And CI has passed on the PR.

kbr- commented:
Maybe there's some problem with the new test suite and the occurence depends on the machine on which it is executed (package versions or something?) Ugh.

kbr- commented:
Ah, there's a conflict with https://github.com/scylladb/scylladb/pull/11482 - the new suite is relying on the toplevel conftest.py (which that PR has removed).

kbr- commented:
I modified the merge commit to resolve the conflict. Let's hope it passes now.
